bokeh/core/has_props.py

```diff
@@ -607,8 +607,17 @@ def apply_theme(self, property_values):
 
         # Emit any change notifications that result
         for k, v in old_values.items():
-            prop = self.lookup(k)
-            prop.trigger_if_changed(self, v)
+            descriptor = self.lookup(k)
+            if isinstance(v, PropertyValueContainer) and v._unmodified_default_value:
+                if k in self._property_values[k]:
+                    self._property_values[k]._unregister_owner(self, descriptor)
+                if k in property_values:
+                    self._property_values[k] = descriptor.property.prepare_value(self.__class__, k, property_values[k])
+                else:
+                    self._property_values[k] = descriptor.class_default(self.__class__)
+                self._property_values[k]._unmodified_default_value = True
+                self._property_values[k]._register_owner(self, descriptor)
+            descriptor.trigger_if_changed(self, v)
```

Comment:
This code uses a lot of private attributes on `PropertyValueContainer` maybe it can be moved to a method on `PropertyValueContainer` but what is a good name for this? 

Also, the since even "getting" for a property container puts the property in `_property_values` (but marking it as "unmodified") the strategy here is: 
* if not unmodified, do nothing (basic operation of theming)
* otherwise, unregister value in `_property_values` from its owner (should also delete from `_property_values`?)
* if its in prop values assed in, prepare a new value (i.e. wrap in prop container) and put *that* in `_property_values`
* otherwise (if we are "un-theming" e.g.) make a new `class_default` (is this the right thing to do? maybe can just delete from `_property_values` instead?)
* do all the re-registering that happens to newly wrapped prop containers.